### PR TITLE
fix: validate OpenID discovered configuration, warn when appropriate

### PR DESCRIPTION
**Pull request by fiam**
end_session_endpoint is not a mandatory field, so it should print a
warning instead of an error. Also, add validation for the required
fields according to the spec.

### CODE REVIEW

File: pkg/apihandler/apihandler.go [hunk 1]
```diff
@@ -2003,12 +2003,13 @@ func (r *Builder) configureOpenIDConnectIssuerLogoutURLs() map[string]string {
 			)
 			continue
 		}
-		if config.EndSessionEndpoint == "" {
-			r.log.Error("failed to get openid-configuration",
-				zap.String("end_session_endpoint", config.EndSessionEndpoint),
-			)
+		if err := config.Validate(); err != nil {
+			r.log.Warn("failed to get openid-configuration", zap.Error(err))
```
Comment (StarpTech) [inline pkg/apihandler/apihandler.go:2007]: No warn, debug. It is optional.

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: pkg/apihandler/apihandler.go (hunk 1 above)
Comment (StarpTech) [reply pkg/apihandler/apihandler.go:2007]: Sry, I mean for the EndSessionEndpoint endpoint

File: pkg/apihandler/apihandler.go [hunk 2]
```diff
@@ -2003,12 +2003,13 @@ func (r *Builder) configureOpenIDConnectIssuerLogoutURLs() map[string]string {
...
 			continue
 		}
+		if config.EndSessionEndpoint == "" {
+			r.log.Warn("issuer doesn't support end_session_endpoint", zap.String("issuer", issuer))
```
Comment (jensneuse) [inline pkg/apihandler/apihandler.go:2011]: Could be info or debug I think.

File: pkg/apihandler/apihandler.go [hunk 3]
```diff
@@ -2024,6 +2025,30 @@ type OpenIDConnectConfiguration struct {
...
+	}
+	if c.AuthorizationEndpoint == "" {
+		return openIDConnectConfigurationMissingFieldError("authorization_endpoint")
+	}
+	if c.TokenEndpoint == "" {
+		// TODO: This might be optional with Implicit Flow?
+		return openIDConnectConfigurationMissingFieldError("token_endpoint")
+	}
+	if c.JwksUri == "" {
+		return openIDConnectConfigurationMissingFieldError("jwks_uri")
```
Comment (jensneuse) [inline pkg/apihandler/apihandler.go:2047]: this is optional I think

File: pkg/apihandler/apihandler.go [hunk 4]
```diff
@@ -2024,6 +2025,30 @@ type OpenIDConnectConfiguration struct {
...
+	// See https://openid.net/specs/openid-connect-discovery-1_0.html
+	if c.Issuer == "" {
+		return openIDConnectConfigurationMissingFieldError("issuer")
+	}
+	if c.AuthorizationEndpoint == "" {
+		return openIDConnectConfigurationMissingFieldError("authorization_endpoint")
+	}
+	if c.TokenEndpoint == "" {
+		// TODO: This might be optional with Implicit Flow?
+		return openIDConnectConfigurationMissingFieldError("token_endpoint")
```
Comment (jensneuse) [inline pkg/apihandler/apihandler.go:2044]: we don't want to support implicit flow, only PKCE

**Review by jensneuse — CHANGES_REQUESTED** [review]

File: pkg/apihandler/apihandler.go (hunk 1 above)
Comment (fiam) [reply pkg/apihandler/apihandler.go:2007]: 👍

Comment (fiam) [reply pkg/apihandler/apihandler.go:2007]: Changed to debug

File: pkg/apihandler/apihandler.go (hunk 4 above)
Comment (fiam) [reply pkg/apihandler/apihandler.go:2044]: Changed to debug

File: pkg/apihandler/apihandler.go (hunk 2 above)
Comment (fiam) [reply pkg/apihandler/apihandler.go]: Changed to debug

File: pkg/apihandler/apihandler.go (hunk 3 above)
Comment (fiam) [reply pkg/apihandler/apihandler.go:2047]: Maybe I’m looking at the wrong spec, but it says it’s required here https://openid.net/specs/openid-connect-discovery-1_0.html

Comment (jensneuse) [reply pkg/apihandler/apihandler.go:2047]: I'm just not sure if it will always be present...

Comment (fiam) [reply pkg/apihandler/apihandler.go]: Made it optional

**Review by jensneuse — APPROVED** [review]

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**